tokio-stream/src/stream_ext/map_while.rs

```diff
@@ -0,0 +1,72 @@
+use crate::Stream;
+
+use core::fmt;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+use pin_project_lite::pin_project;
+
+pin_project! {
+    /// Stream for the [`map_while`](super::StreamExt::map_while) method.
+    #[must_use = "streams do nothing unless polled"]
+    pub struct MapWhile<St, F> {
+        #[pin]
+        stream: St,
+        f: F,
+        done: bool
+    }
+}
+
+impl<St, F> fmt::Debug for MapWhile<St, F>
+where
+    St: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MapWhile")
+            .field("stream", &self.stream)
+            .finish()
+    }
+}
+
+impl<St, F> MapWhile<St, F> {
+    pub(super) fn new(stream: St, f: F) -> Self {
+        MapWhile {
+            stream,
+            f,
+            done: false,
+        }
+    }
+}
+
+impl<St, F, T> Stream for MapWhile<St, F>
+where
+    St: Stream,
+    F: FnMut(St::Item) -> Option<T>,
+{
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T>> {
+        if !*self.as_mut().project().done {
+            self.as_mut().project().stream.poll_next(cx).map(|ready| {
+                let ready = ready.and_then(self.as_mut().project().f);
+
+                if ready.is_none() {
+                    *self.as_mut().project().done = true;
+                }
+
+                ready
+            })
+        } else {
+            Poll::Ready(None)
+        }
```

Comment:
The code here is quite noisy with all the `self.as_mut().project()` repetitions. Consider something like this:
```suggestion
        let me = self.project();
        match me.stream.as_pin_mut() {
            Some(stream) => match stream.poll_next(cx).map(|opt| opt.and_then(me.f)) {
                Poll::Ready(Some(item)) => Poll::Ready(Some(item)),
                Poll::Ready(None) => {
                    me.stream.set(None);
                    Poll::Ready(None)
                },
                Poll::Pending => Poll::Pending,
            },
            None => Poll::Ready(None),
        }
```

Reply:
I agree with reducing the noise in the implemetnation, but I'm not sure this will work. In particular, `me.stream.as_pin_mut` will consume the `Pin<&mut Option<St>>` and then we won't be able to call `set` on it. I can't find any other way of accessing the `Pin<&mut St>` we need to call `poll_next` on the underlying structure.

On the other hand, if we leave `St` as it is but then make the `F` optional, (`f: Option<F>`), because pinning is not structural to `f`, we can empty out the `Option<F>` whenever we want. I am not sure if the `Option` optimization is less likely to apply to `F` than it is to the `St` though. For what it's worth, the `skip_while` combinator also makes the associated closure optional, to the `St`.

Would it make sense to make the `F` optional instead of the `St`? 

I'm not super comfortable with pinning so maybe there's something obvious I'm missing.

Reply:
You can do `me.stream.as_mut().as_pin_mut()`.

Reply:
!! Thanks, that's useful.